tutorial/tests/testsuite_helpers.py

```diff
@@ -257,24 +323,45 @@ class ResultCollector:
     def __init__(self) -> None:
         self.tests: Dict[str, TestResult] = {}
 
+    def pytest_runtest_makereport(self, item: pytest.Item, call: pytest.CallInfo):
+        if call.when == "call":
+            if call.excinfo is None:
+                self.tests[item.nodeid] = TestResult(
+                    test_name=item.nodeid,
+                    success=True,
+                    stdout=call.result,
+                    stderr=call.result,
+                    exception=None,
+                    traceback=None,
+                )
+            else:
+                self.tests[item.nodeid] = TestResult(
+                    test_name=item.nodeid,
+                    success=False,
+                    stdout=None,
+                    stderr=None,
+                    exception=traceback.format_exception_only(call.excinfo.value),
```

Comment:
I would prefer that the formatting happens at the time of construction of the test output. 
We could store the excinfo here and call the method later.